### PR TITLE
allow to merge forks

### PR DESCRIPTION
**Pull request by amtoine**
- should fix https://github.com/amtoine/nu-git-manager/issues/52
- :warning: ~~wait for #48 to have the end-to-end tests~~

## description
- add a new `nu-git-manager git repo` module with
  - `get-root-commit` to get the first commit of a local repo
  - `is-grafted` to check for [*grafted* repos](https://stackoverflow.com/questions/27296188/what-exactly-is-a-grafted-commit-in-a-shallow-clone)
  - the two commands are tested in `tests root-commit` and `tests detect-grafting` respectively
  - in order for them to work, Git is now configured in the CI
- add a `throw-warning` command
- store `record<path: path, grafted: bool, root: string>` in the cache to detect forks
  - update the tests and the cache updates in `gm`
- detect forks when cloning a new repo and give a warning without changing anything
 
### what it looks like to clone repos now
- cloning a *grafted* repo, i.e. which won't be used for fork detection
```nushell
> tk run --clean { gm clone https://github.com/amtoine/nushell --depth 1 }
Cloning into '/tmp/foo/github.com/amtoine/nushell'...
...
Warning:   ⚠ cloning_grafted_repository
| this repo is grafted, cannot detect forks
updating cache... done
```
- cloning a first fork of Nushell
```nushell
> tk run --clean { gm clone https://github.com/fdncred/nushell }
Cloning into '/tmp/foo/github.com/fdncred/nushell'...
...
updating cache... done
```
- cloning a second fork of Nushell: one fork detected
```nushell
> tk run { gm clone https://github.com/kubouch/nushell }
Cloning into '/tmp/foo/github.com/kubouch/nushell'...
...
Warning:   ⚠ cloning_fork
| this repo is a fork of 1 other repo because they share the same root commit: 8f3b273337b53bd86d5594d5edc9d4ad7242bd4c
| - github.com/fdncred/nushell
updating cache... done
```
- cloning a last fork of Nushell: two forks detected
```nushell
> tk run { gm clone https://github.com/nushell/nushell }
Cloning into '/tmp/foo/github.com/nushell/nushell'...
...
Warning:   ⚠ cloning_fork
| this repo is a fork of 2 other repos because they share the same root commit: 8f3b273337b53bd86d5594d5edc9d4ad7242bd4c
| - github.com/fdncred/nushell
| - github.com/kubouch/nushell
updating cache... done
```
> **Note**
> the real rendering:
> 2023-11-06-121223_1920x1080_scrot

## TODO
- [x] add root commit and graf status to the cache
- [x] give a warning when cloning a non-grafted fork
- [x] add a command to merge forks together
- [x] do the TODOs and FIXMEs in the changes

### CODE REVIEW

**Comment by melMass** [comment]
Nice, I wouldn't expose "get-root-commit" and in the cache I would change "root" to "hash"

File: src/nu-git-manager/git/repo.nu [hunk 1]
```diff
@@ -0,0 +1,25 @@
+use ../error/error.nu [throw-error]
+
+# tell if a local repository has been grafted
+#
+# > see more about grafted commits
+# > [here](https://stackoverflow.com/questions/27296188/what-exactly-is-a-grafted-commit-in-a-shallow-clone)
+export def is-grafted [
+    repo?: path, # the path to the repository to check (defaults to `pwd`)
+]: nothing -> bool {
+    let repo = $repo | default (pwd)
```
[the remark this answers is not in the document]
Comment (amtoine) [reply src/nu-git-manager/git/repo.nu:10]: nope we don't, which is great for simple oneliners like this or even longer pipelines that can be written as
```nushell
let foo = $bar
    | foo
    | bar
    | baz
    | brrr
```
:smirk:

**Comment by amtoine** [comment]
> Nice, I wouldn't expose "get-root-commit" (or maybe in sugar)
>
`get-root-commit` is required in `mod.nu` so we need to `export` it :thinking: 

> in the cache I would change "root" to "hash"
>
i've renamed it to `root_hash` in 14a2f69, to keep the idea that it's the hash of the root commit, it's not the *hash* of the files for instance :yum:

File: src/nu-git-manager/git/repo.nu (hunk 1 above)
[the remark this answers is not in the document]
Comment (melMass) [reply src/nu-git-manager/git/repo.nu:10]: That was still possible with parentheses. It used to be without, then parentheses were enforced and now it seems to be lifted again

**Comment by amtoine** [comment]
that Windows CI was not that bad :relieved:

**Pull request merged**